lib/src/room.dart

```diff
@@ -1648,102 +1650,171 @@ class Room {
+          ),
+          client,
         );
-        _requestingMatrixIds.remove(mxID);
-
-        // Store user in database:
-        await client.database?.transaction(() async {
-          await client.database?.storeEventUpdate(
-            EventUpdate(
-              content: foundUser!.toJson(),
-              roomID: id,
-              type: EventUpdateType.state,
-            ),
-            client,
-          );
-        });
-      } on MatrixException catch (_) {
-        // Ignore if we have no permission
-      } catch (e, s) {
-        if (!ignoreErrors) {
-          _requestingMatrixIds.remove(mxID);
-          rethrow;
-        } else {
-          Logs().w('Unable to request the user $mxID from the server', e, s);
-        }
+      });
+
+      return foundUser;
+    } on MatrixException catch (_) {
+      // Ignore if we have no permission
+      return null;
+    } catch (e, s) {
+      if (!ignoreErrors) {
+        rethrow;
+      } else {
+        Logs().w('Unable to request the user $mxID from the server', e, s);
+        return null;
       }
     }
+  }
 
-    // User not found anywhere? Set a blank one:
-    foundUser ??= User(mxID, room: this, membership: 'leave');
-
-    // Is it a left user without any displayname/avatar info? Try fetch profile:
-    if (requestProfile &&
-        {Membership.ban, Membership.leave}.contains(foundUser.membership) &&
-        foundUser.displayName == null &&
-        foundUser.avatarUrl == null) {
-      try {
-        final profile = await client.getProfileFromUserId(mxID);
-        foundUser = User(
-          mxID,
-          displayName: profile.displayName,
-          avatarUrl: profile.avatarUrl?.toString(),
-          membership: Membership.leave.name,
-          room: this,
-        );
-      } catch (e, s) {
-        if (!ignoreErrors) {
-          rethrow;
-        } else {
-          Logs().w('Unable to request the profile $mxID from the server', e, s);
+  // Internal helper to implement requestUser
+  Future<User?> _requestUser(
+    String mxID, {
+    required bool ignoreErrors,
+    required bool requestState,
+    required bool requestProfile,
+  }) async {
+    // Is user already in cache?
+    final userFromState = getState(EventTypes.RoomMember, mxID)?.asUser(this);
+
+    // If not in cache, try the database
+    var foundUser = userFromState;
+
+    // If the room is not postloaded, check the database
+    if (partial && foundUser == null) {
+      foundUser = await client.database?.getUser(mxID, this);
+    }
+
+    // If not in the database, try fetching the member from the server
+    if (requestState && foundUser == null) {
+      foundUser = await _requestSingleParticipantViaState(mxID,
+          ignoreErrors: ignoreErrors);
+    }
+
+    // If the user isn't found or they have left and no displayname set anymore, request their profile from the server
+    if (requestProfile) {
+      if (foundUser
+          case null ||
+              User(
```

Comment:
Is it possible to have a foundUser with membership leave/ban but displayName not null here?

Reply:
Possibly. Synapse unsets the displayname on leaves, but I don't think it has to be that way and nothing in the spec suggests it would.

Reply:
Hm displayName isn't checked in the == override anyway, can we just remove it then?

Reply:
What == override? The original requirement was to load the profile, if we have no displayname (or avatar) set on a leave event, so that's what this does. If we always load the profile, then we would also fetch it, if we already fetched the displayname previously.

Reply:
Makes sense, but the User object == override does not take the displayname and avatarUrl into account. So are you sure this still works as expected?